## Step 1:
- Pandas library was imported first.
- read_csv() used to convert the dataset into a DataFrame.
- A total of 891 rows and 12 columns are availabe.

In [49]:
#step1
import pandas as pd
df= pd.read_csv("train.csv")

print(df.shape)

(891, 12)


In [50]:
columns = df.columns

print(columns)

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')


"columns" used to list the dataset column names and print them out.

In [51]:
print(df.dtypes)

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object


"dtypes" used to list the data type of each column.

## Step 2: 

In [52]:
#step2 
missing = df.isnull().sum()

print(missing)

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


Rows with missing values are listed above.

In [53]:
df_new = df.drop(columns=["Cabin"])
print(df_new)

     PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ... 

The DataFrame is modified by dropping the "Cabin" column - has a lot of missing rows- because deleting the missing rows entirely will remove most of the table data.

In [55]:
x = df_new["Age"].median()
print("Average age:", x)
df_new["Age"]= df_new["Age"].fillna(x)
print(df_new)

Average age: 28.0
     PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 

"Age" column, in contrast, is filled on the missing rows with the average age for efficiency.

In [29]:
df_new = df_new.dropna(subset=["Embarked"])
print(df_new)

     PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ... 

Rows with missing data in "Embarked" column are dropped (minimal impact), bringing the table to a total of 889 rows and 11 columns.

In [30]:
df_new = df_new.drop_duplicates()
print(df_new.shape)

(889, 11)


Since there is no duplicates, the number of rows remains the same.

## Step 3:

In [36]:
#step3
df_avg_fare = df_new["Fare"].mean()
print(df_avg_fare)

32.09668087739032


The average Fare calculated using mean() and printed.

In [37]:
high_fare = df_new[df_new["Fare"]>df_avg_fare]
print(high_fare.shape[0], "out of", df_new.shape[0],"passengers paid above the average!")
print(high_fare)


209 out of 889 passengers paid above the average!
     PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
6              7         0       1   
23            24         1       1   
27            28         0       1   
..           ...       ...     ...   
856          857         1       1   
863          864         0       3   
867          868         0       1   
871          872         1       1   
879          880         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
23                        Sloper, Mr. William Thompson    male  28.0      0   
27                      Fortune, Mr. Charles Alexander    male  19.0      3   
..                 

Fares higher than the average are filterd with boolean masking and printed.

### Findings:
  1- The average Fare for is 32.09
  2- Fares higher than the average are 209 out of 889 (approximately quarter the total Fares).
  3- Mjority of passengers who paid above the average are first class passengers.

# Step 4:
- "groupby" and "agg()"used to group data to display survival rate, Age and Fare per class.
- It was also used to calculate the survival rate based on Gender.

In [41]:
#step4
class_based = df_new.groupby("Pclass").agg({"PassengerId" : "count",
                             "Survived" : "mean",
                             "Age" : "mean",
                             "Fare" : "mean"}
                            )
print(class_based)


        PassengerId  Survived        Age       Fare
Pclass                                             
1               214  0.626168  36.688879  84.193516
2               184  0.472826  29.765380  20.662183
3               491  0.242363  25.932627  13.675550


Fisrt class passengers have the highest survival rate while paying the highest Fares in Average.
Third class passengers form a majority of the train passengers and they form the youngest group of passengers in average.

In [42]:
gender_based = df_new.groupby("Sex")["Survived"].mean()   
print(gender_based)

Sex
female    0.740385
male      0.188908
Name: Survived, dtype: float64


Most survivors of the train are females.